Methods to downgrade from 2020 -> 2018 / 2015

MadhuvanthG opened the pull request:
- Unlike the existing downgrade/upgrade functions which are part of a separate package, for 2020 version structs, these schema upgrade/downgrade are methods on the types itself

- Because 2020 spec only has additions, used a simple approach of unmarshaling the  2020 object directly into 2018, which would get rid of the additional elements in 2020. But with this approach, (2020)ViewingPolicy.Allocation gets unmarshaled  into (2018)ViewingPolicy.ActionProperty. So explicitly removing it in the downgraded version

- **Will add unit tests**, but want to first gather feedback on the approach

superflaco commented on README.md:
```diff
@@ -2,12 +2,15 @@
+2018 schema: http://www.scte.org/schemas/224/SCTE224-20180501.xsd (link no longer working, refer to the spec located alongside structs in this repo)
 
-2018 schema: http://www.scte.org/schemas/224/SCTE224-20180501.xsd
+Latest schema (currently points to 2020): https://www.scte.org/standards-development/library/standards-catalog/ansiscte-224-2018r1/
```
wow, super weird that the latest schema link https://scte-cms-resource-storage.s3.amazonaws.com/SCTE_224_20200407-1600425639358.xsd is found on a page referencing the 2021 schema located on a url mentioning 2018r1 @skurkows seems like SCTE still has wacky page structures

superflaco commented on types/scte224v20200407/sctestructures.go:
```diff
@@ -42,6 +46,46 @@ type Media struct {
+
+	// Unmarshal into 2018 struct
+	var media2018 *scte224_2018.Media
+	err = xml.Unmarshal(encodedMedia, &media2018)
```
how does this benchmark?  I think I avoided a marshal-unmarshal route with 2015-2018 conversions both due to the namespace difference as well as thinking marshalling performance would be terrible.  Might not be a big deal though mostly curious.

MadhuvanthG replied:
Valid concern with Unmarshaling/marshaling. I am skeptical about it too. 

Yet to benchmark, was going to do so while adding unit tests.

We could leave the PR open, until I benchmark and observe results.

MadhuvanthG replied:
We still have to dereference each of the values, if we were to manually copy from one struct to another. 

So I thought, at least theoretically, there wouldn’t be too much difference between marshaling/unmarshaling approach and copying approach.

superflaco commented on types/scte224v20200407/viewingpolicy.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+//Table 12
+type ViewingPolicy struct {
```
got too cumbersome in one big file?

MadhuvanthG replied:
Yes, was getting bigger with the new methods

superflaco reviewed (approved):
Seems fine, but I just get squeamish about all the marshalling roundtrips killing performance which may be overthinking it.  Let's get the readme cleaned up before committing though.

MadhuvanthG commented:
> Seems fine, but I just get squeamish about all the marshalling roundtrips killing performance which may be overthinking it. Let's get the readme cleaned up before committing though.

Marshaling/Unmarshaling wasn't helping performance at all. I did some benchmarks and for a Media with ~1000 media points:
1. xml marshal/unmarshal approach took ~400 milliseconds
2. field-by-field copy took under 5 milliseconds

Although approach (1) leads to simpler code, that's a huge performance penalty to pay. So, changed the methods to use approach (2) and added enough unit tests to ensure downgrade of different structs are working as expected.
